### PR TITLE
Update alzDefaultPolicyAssignments.bicep

### PR DESCRIPTION
**Pull request by VeronicaSea**
# Overview/Summary

Those changes are used to have ALZ modules should be deployed optionally based on the SLZ flag to control the policy enforcement. All the SLZ modules should use the SLZ flag for SLZ policy enforcement or make it ‘Default’.  Besides, the SLZ policy assignment modules supports policy Effect param, which we previously supported.

## This PR fixes/adds/changes/removes

Issue 1 : Currently, all the ALZ modules will be deployed when the policy assignment file is invoked. ALZ modules should be deployed optionally based on the flag. 

Issue 2 : The SLZ policy assignment modules are using ALZ flag to control the policy enforcement. All the SLZ modules should use the SLZ flag for SLZ policy enforcement or make it ‘Default’. 

Issue 3 : The SLZ policy assignment modules don’t have policy Effect param, which we previously supported.

### Breaking Changes

N/A

## Testing Evidence

Replace this with any testing evidence to show that your Pull Request works/fixes as described and planned (include screenshots, if appropriate).

## As part of this Pull Request I have

- [x] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [x] Ensured the resource API versions in `.bicep` file/s I am adding/editing are using the latest API version possible
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [ ] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [ ] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [ ] Performed testing and provided evidence.
- [ ] Updated one or more of the following tests *(if required)*
  - [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml)
  - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows)
  - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
  - [ValidateAzCloud (Base validation in Azure Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/base-unit-validate.yml)
  - [ValidateMcCloud (Base validation in Azure China Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/mc-base-unit-validate.yml)
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep [hunk 1]
```diff
@@ -101,6 +101,9 @@ param parExcludedPolicyAssignments array = []
 @sys.description('Set Parameter to true to Opt-out of deployment telemetry')
 param parTelemetryOptOut bool = false
 
+@sys.description('Effect type for all policy definitions')
+param parPolicyEffect string = 'Deny'
```
Comment (azure-lm) [inline infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep:105]: I feel this name is too vague and could be interpreted in a variety of ways. Perhaps we need to use the 'sovereignty' keyword somewhere in the variable name.

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/generateddocs/alzDefaultPolicyAssignments.bicep.md [hunk 2]
```diff
@@ -182,6 +183,14 @@ Set Parameter to true to Opt-out of deployment telemetry
 
 - Default value: `False`
 
+### parPolicyEffect
+
+![Parameter Setting](https://img.shields.io/badge/parameter-optional-green?style=flat-square)
+
+Effect type for all policy definitions
```
Comment (azure-lm) [inline infra-as-code/bicep/modules/policy/assignments/alzDefaults/generateddocs/alzDefaultPolicyAssignments.bicep.md:190]: Likewise, this isn't the effect for all policies, just the sov baseline.

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep (hunk 1 above)
Comment (VeronicaSea) [reply infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep]: Updated with Sovereignty.

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/generateddocs/alzDefaultPolicyAssignments.bicep.md (hunk 2 above)
Comment (VeronicaSea) [reply infra-as-code/bicep/modules/policy/assignments/alzDefaults/generateddocs/alzDefaultPolicyAssignments.bicep.md]: Resolved to have sovereign policies in description.

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep [hunk 3]
```diff
@@ -101,6 +101,9 @@ param parExcludedPolicyAssignments array = []
 @sys.description('Set Parameter to true to Opt-out of deployment telemetry')
 param parTelemetryOptOut bool = false
 
+@sys.description('Effect type for sovereign policy definitions')
+param parSovereigntyPolicyEffect string = 'Deny'
```
Comment (oZakari) [inline infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep:105]: Can we please add this params `parPolicyAssignmentSovereigntyConfidential` and `parTopLevelPolicyAssignmentSovereigntyGlobal` user-defined type objects as properties and only allow valid policy effects? I think it is a bit cleaner to keep the sovereign stuff intertwined as much as possible and then also gives the ability to set different effects per assignment.

**Review by oZakari — CHANGES_REQUESTED** [review]
Hey @vericasea, once you incorporate the policy effect param into the UDT properties, can you also please add the new properties to the params within this parameters [file](https://github.com/Azure/ALZ-Bicep/blob/main/infra-as-code/bicep/modules/policy/assignments/alzDefaults/parameters/alzDefaultPolicyAssignments.parameters.all.json)?

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep [hunk 4]
```diff
@@ -1589,9 +1598,12 @@ module modPolicyAssignmentLzsConfidentialCorpEnforceSovereigntyConf '../../../po
       allowedVirtualMachineSKUs: {
         value: !(empty(parPolicyAssignmentSovereigntyConfidential.parAllowedVirtualMachineSKUs)) ? parPolicyAssignmentSovereigntyConfidential.parAllowedVirtualMachineSKUs : varPolicyAssignmentEnforceSovereignConf.libDefinition.properties.parameters.allowedVirtualMachineSKUs.value
       }
+      effect: {
+        value: parSovereigntyPolicyEffect
+      }
     }
     parPolicyAssignmentIdentityType: varPolicyAssignmentEnforceSovereignConf.libDefinition.identity.type
-    parPolicyAssignmentEnforcementMode: parDisableAlzDefaultPolicies ? 'DoNotEnforce' : varPolicyAssignmentEnforceSovereignConf.libDefinition.properties.enforcementMode
+    parPolicyAssignmentEnforcementMode: parLandingZoneMgConfidentialEnable ? 'Default' : varPolicyAssignmentEnforceSovereignConf.libDefinition.properties.enforcementMode
```
Comment (oZakari) [inline infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep:1606]: I looked back at these policy assignments (here is one for [example](https://github.com/Azure/ALZ-Bicep/blob/main/infra-as-code/bicep/modules/policy/assignments/lib/policy_assignments/policy_assignment_es_enforce_sovereignty_baseline_conf.tmpl.json): , and this is redundant as the default value within the assignment is Default.

**Comment by VeronicaSea** [comment]
> Hey @vericasea, once you incorporate the policy effect param into the UDT properties, can you also please add the new properties to the params within this parameters [file](https://github.com/Azure/ALZ-Bicep/blob/main/infra-as-code/bicep/modules/policy/assignments/alzDefaults/parameters/alzDefaultPolicyAssignments.parameters.all.json)?

Thanks and added them in the PR.

**Comment by oZakari** [comment]
> > Hey @vericasea, once you incorporate the policy effect param into the UDT properties, can you also please add the new properties to the params within this parameters [file](https://github.com/Azure/ALZ-Bicep/blob/main/infra-as-code/bicep/modules/policy/assignments/alzDefaults/parameters/alzDefaultPolicyAssignments.parameters.all.json)?
> 
> Thanks and added them in the PR.

Thanks @VeronicaSea, I've moved the params into the UDT for each param and added them to the parameters file.

I think the param to disable the SLZ policies is probably easier to understand if left out. You can still disable the deployment of the policy entirely with `parTopLevelSovereigntyGlobalPoliciesEnable` for the global policies or `parPolicyAssignmentSovereigntyConfidential` for the confidential sovereign policies. Now there is also the capability to add the audit effect which is essentially the same thing. Just think it's a bit cleaner, but let me know if you want to discuss further.

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep [hunk 5]
```diff
@@ -89,6 +99,9 @@ param parPrivateDnsZonesNamesToAuditInCorp array = []
 @sys.description('Set Enforcement Mode of all default Policies assignments to Do Not Enforce.')
 param parDisableAlzDefaultPolicies bool = false
 
+@sys.description('Set Enforcement Mode of all default sovereign Policies assignments to Do Not Enforce.')
+param parDisableSlzDefaultPolicies bool = false
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep:103]: This is perfect 👍

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep [hunk 6]
```diff
@@ -7,6 +7,9 @@ type policyAssignmentSovereigntyGlobalOptionsType = {
 
   @sys.description('The list of locations that your organization can use to restrict deploying resources to. If left empty, only the deployment location will be allowed.')
   parListOfAllowedLocations: string[]
+
+  @sys.description('The effect type for the Sovereignty Baseline - Global Policies Assignment.')
+  parPolicyEffect: ('Audit' | 'Deny' | 'Disabled' | 'AuditIfNotExists')
```
Comment (jtracey93) [inline infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep:12]: This is to broad as a parameter name and will cause confusion also seems a duplicate of `parTopLevelSovereignGlobalPoliciesEnable`?

Please can we align and make this simplified and clear to customers what the parameter does

**Review by jtracey93 — CHANGES_REQUESTED** [review]

File: infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep (hunk 6 above)
Comment (VeronicaSea) [reply infra-as-code/bicep/modules/policy/assignments/alzDefaults/alzDefaultPolicyAssignments.bicep:12]: @oZakari How do you think?

**Pull request merged**